com.unity.sg2/Editor/GraphDeltaRegistry/PreviewManager/PreviewService.cs

```diff
@@ -422,6 +416,74 @@ public List<string> NotifyNodeFlowChanged(string nodeName, bool wasNodeDeleted =
                 ExecuteLater(0);
         }
 
+        // ==================================
+        // Basic Async Implementation
+
+        /// <summary>
+        /// UpdateHandler is registered with EditorApplication.update in Initialize.
+        /// It is called about every m_updateThresholdSeconds seconds.
+        /// </summary>
+        void UpdateHandler()
+        {
+            var curTimeSeconds = EditorApplication.timeSinceStartup;
+            var deltaTimeSeconds = curTimeSeconds - m_LastUpdateTimeSeconds;
+            if (deltaTimeSeconds < m_UpdateThresholdSeconds)
+                return;
+            // do update
+            UpdateShaderCompilationStatus();
+            // change last updated
+            m_LastUpdateTimeSeconds = curTimeSeconds;
+        }
+
+        /// <summary>
+        /// Checks the list of preview data that are currently being compiled for
+        /// completion.
+        /// Upon compile completion:
+        ///   update material textures
+        ///   check for errors
+        ///   mark dirty
+        /// </summary>
+        void UpdateShaderCompilationStatus()
+        {
+            if (m_PreviewsCompiling.Count <= 0)
+                return;
+            List<PreviewData> completed = new List<PreviewData>();
+            foreach (PreviewData previewToUpdate in m_PreviewsCompiling)
+            {
+                bool isCompiling = false;
+                for (int i = 0; i < previewToUpdate.material.passCount; ++i)
+                {
+                    if (!ShaderUtil.IsPassCompiled(previewToUpdate.material, i))
+                    {
+                        isCompiling = true;
+                        break;
+                    }
+                }
+                if (!isCompiling)
+                {
+                    // add to the local completed list for update
+                    completed.Add(previewToUpdate);
+
+                    // set the material textures
+                    foreach (var texDefault in previewToUpdate.defaultTextures)
+                        previewToUpdate.material.SetTexture(texDefault.Item1, texDefault.Item2);
+
+                    // check for PreviewData errors
+                    if (CheckForErrors(previewToUpdate))
+                        previewToUpdate.hasShaderError = true;
+
+                    // mark dirty
+                    previewToUpdate.isShaderOutOfDate = false;
+                }
+            }
+            foreach (PreviewData completedPreview in completed)
+            {
+                m_PreviewsCompiling.Remove(completedPreview);
+            }
+        }
+
+        // END Basic Async Implementation
+        // ==================================
```

Comment:
Just verifying-- the PreviewTaskRunner and CreateShaderTask are not in use yet, and the above is currently what's driving the async?

The goal is to replace each individual step here with IPreviewTasks that are chained together as necessary to satisfy either a given request and/or for pushing changes to the PreviewUpdateListener?

Reply:
Yes, that's my concept.